Data/Map/Lazy.hs

```diff
@@ -14,19 +14,66 @@
 -- Maintainer  :  libraries@haskell.org
 -- Portability :  portable
 --
--- An efficient implementation of ordered maps from keys to values
--- (dictionaries).
 --
--- API of this module is strict in the keys, but lazy in the values.
--- If you need value-strict maps, use "Data.Map.Strict" instead.
--- The 'Map' type itself is shared between the lazy and strict modules,
--- meaning that the same 'Map' value can be passed to functions in
--- both modules (although that is rarely needed).
+-- = Lazy Maps (aka dictionaries)
 --
--- These modules are intended to be imported qualified, to avoid name
--- clashes with Prelude functions, e.g.
+-- The @'Map' k v@ type represents an /ordered/ map indexed by keys of type @k@
+-- containing /lazy/ values of type @v@.
 --
--- >  import qualified Data.Map.Lazy as Map
+-- When deciding if this is the correct data structure to use, consider:
+--
+-- * The API of this module is strict in the keys but lazy in the values. If you
+-- don't need /value-lazy/ maps, use "Data.Map.Strict" instead.
+--
+-- * If your key type is @Int@, use "Data.IntMap.Lazy" instead.
+--
+-- * If you don't care about ordering, use @Data.HashMap.Lazy@ from the
+-- <https://hackage.haskell.org/package/unordered-containers unordered-containers>
+-- package instead.
+--
+-- For a walkthrough of the most commonly used functions see the
+-- <https://haskell-containers.readthedocs.io/en/latest/map.html maps introduction>.
+--
+-- This module is intended to be imported qualified, to avoid name clashes with
+-- Prelude functions:
+--
+-- > import qualified Data.Map.Lazy as Map
+--
+-- Note that the implementation is /left-biased/, the elements of a first
+-- argument are always preferred to the second, for example in 'union' or
+-- 'insert'.
+--
+--
+-- == Detailed performance information
+--
+-- The amortized running time is given for each operation, with /n/ referring to
+-- the number of entries in the map.
+--
+-- Benchmarks comparing "Data.Map.Lazy" with other dictionary implementations
+-- can be found at https://github.com/haskell-perf/dictionaries.
+--
+--
+-- == Strictness propertios
+--
+-- This module satisfies the following strictness property:
+--
+-- * Key arguments are evaluated to WHNF
+--
+-- Here are some examples that illustrate the property:
+--
+-- > insertWith (\ new old -> old) undefined v m  ==  undefined
+-- > insertWith (\ new old -> old) k undefined m  ==  OK
+-- > delete undefined m  ==  undefined
```

Comment:
This `delete` example isn't even an example (Can you imagine *any* sort of finite map implementation with a lazy `delete`? What information could you actually obtain from a map from which an unknown and possibly-absent key has been deleted?) But I think all these examples should go away.